### ISSUE TITLE
Issue using custom generator at validation time (fit_generator)

### ISSUE THREAD

**Issue by mgendarme** [rstudio/keras#1215]
Hello the R deep learning community, 
It looks like I have to re-open [this issue](https://github.com/rstudio/keras/pull/1202#issue-599938073) related to #1201.
I just upgrade to keras 2.4.3 and tensorflow 2.4.1.
After doing this and testing a standard model of mine (Feature Pyramid Network with ResNet101 as pre-trained encoder) on a semantic segmentation task it looks like the behaviour of the model at validation time is incorrect, though training is behaving exactly as prior to upgrade. 

```
train_generator <- custom_generator(data = train_input,
                                    shuffle = TRUE,
                                    intensity_operation = FALSE,
                                    batch_size = BATCH_SIZE)

val_generator <- custom_generator(data = val_input,
                                  shuffle = FALSE,
                                  intensity_operation = FALSE,
                                  batch_size = BATCH_SIZE)

history <- model %>% 
  fit_generator(
    generator = train_generator,
    epochs = EPOCHS,
    steps_per_epoch = STEPS_PER_EPOCHS,
    validation_data = val_generator,
    validation_steps = VALIDATION_STEPS,
    verbose = 1L
  )
```
Here is what I get after the model ran for 2 epochs (for testing):
```
2021-04-19 09:48:42.441441: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2021-04-19 09:48:42.460756: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 3691590000 Hz
Epoch 1/2
2021-04-19 09:49:06.403905: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-04-19 09:49:09.296115: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-04-19 09:49:09.953871: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
WARNING:tensorflow:Callback method `on_train_batch_begin` is slow compared to the batch time (batch time: 0.2387s vs `on_train_batch_begin` time: 0.6364s). Check your callbacks.
200/200 - 222s - loss: 0.4489 - dice_coef_loss_for_bce: 0.3109 - val_loss: 0.8434 - val_dice_coef_loss_for_bce: 0.9920
Epoch 2/2
200/200 - 184s - loss: 0.3986 - dice_coef_loss_for_bce: 0.2564 - val_loss: 1.1392 - val_dice_coef_loss_for_bce: 0.9992
Warning message:
In fit_generator(., generator = train_generator, epochs = EPOCHS,  :
  `fit_generator` is deprecated. Use `fit` instead, it now accept generators.
```
Before update `dice_coef_loss_for_bce` would start around 0.3 (like we observe here, on that front everything alright) and `val_dice_coef_loss_for_bce` would be relatively similar to `dice_coef_loss_for_bce` and not close to 1. I think based on previous issues that there is an issue with the generator or how the output of the generator is handle by keras at evaluation time.

I also do not really understand this part of the error message `Callback method on_train_batch_begin is slow compared to the batch time (batch time: 0.2387s vs on_train_batch_begin time: 0.6364s)`. Check your callbacks.` What are we facing here exactly?

For sake of completeness please find down here the type of custom generator I am using:
```
custom_generator <- function(data,
                             shuffle,
                             intensity_operation, 
                             batch_size) {
 
  i <- 1
  function() {
    if(shuffle) {
      indices <- sample(1:nrow(data), size = batch_size)
    } else { 
      if (i + batch_size >= nrow(data) ) 
        i <<- 1
      indices <- c(i:min(i + batch_size - 1, nrow(data)))
      i <<- i + length(indices)
    }
    
    # generate random batch from input data
    input_iter <- data[indices,]
    
    param_CgB_iter <- list()
    param_RotFlipFlopCrop_iter <- list()
    
    for(i in 1:nrow(input_iter)){
      param_CgB_iter <- c(list(param_generator_CgB()), param_CgB_iter)
      param_RotFlipFlopCrop_iter <- c(list(param_generator_RotFlipFlopCropfix()), param_RotFlipFlopCrop_iter)
    }
    
    if(intensity_operation == T){
      input_iter <- input_iter %>%
        mutate(Parameter_CgB = param_CgB_iter,
               Parameter_RotFlipFlopCropfix = param_RotFlipFlopCrop_iter)
      input_iter <- input_iter %>% 
        mutate(Xaug = map2(X, Parameter_CgB, wrapper_random_param, random_CgB),
               Xaug = map2(Xaug, Parameter_RotFlipFlopCropfix, wrapper_random_param, random_RotFlipFlopCropfix),
               Xaug = map(Xaug, set_dimension, 1),
               Yaug = map2(Y, Parameter_RotFlipFlopCropfix, wrapper_random_param, random_RotFlipFlopCropfix)
               ) %>% 
        select(Xaug, Yaug)
    }  else {
    input_iter <- input_iter %>%
      mutate(Parameter_CgB = param_CgB_iter,
             Parameter_RotFlipFlopCropfix = param_RotFlipFlopCrop_iter)
    input_iter <- input_iter %>% 
      mutate(CHANNELS = CHANNELS,
             Xaug = map2(X, Parameter_RotFlipFlopCropfix, wrapper_random_param, random_RotFlipFlopCropfix),
             Xaug = map_if(Xaug, CHANNELS == 1L, set_dimension, 1L), # might not be necessary
             Yaug = map2(Y, Parameter_RotFlipFlopCropfix, wrapper_random_param, random_RotFlipFlopCropfix)
             ) %>% 
      select(Xaug, Yaug)
    select(X, Y, Xaug, Yaug) # for testing
    }
    if(intensity_operation == T){
      input_iter <- input_iter %>% 
        mutate(Xaug = map2(Xaug, Parameter_CgB, wrapper_random_param, random_CgB))
    }

    X <- list2tensor(input_iter$Xaug)
    Y <- list2tensor(input_iter$Yaug)
    
    list(X, Y)
    
  }
}
```
Any idea where the problem is coming from and how I could fix it?

Thanks a lot.

_Originally posted by @mgendarme in https://github.com/rstudio/keras/issues/1202#issuecomment-823089822_

**Comment by mgendarme** [comment]
Just to make sure this is a weird behaviour from the fit_generator I fed my train_input data to both the `train_generator` and `val_generator` and got the exact same outcome. At validation time the values are still very high.

**Comment by mgendarme** [comment]
I finally managed to solve this issue.
When using pre-trained encoder trained on "imagenet" one has to use `imagenet_preprocess_input` with tensorflow 2.4.1 (haven't tried older version). My data was normalized prior training but this was not sufficient to fix this issue. The generator was not at fault despite only getting issues at validation time.
I must say this is a very strange behaviour. Is it possible that in version older than tensorflow 2.0 some preprocessing was done when using a model trained on Imagenet? I haven't found the answer yet but would be interested to know.
